app/assets/javascripts/index/directions.js

```diff
@@ -111,6 +113,20 @@ OSM.Directions = function (map) {
     return endpoint;
   }
 
+  $(".directions_form .reverse_directions").on("click", function() {
+    var input_from = endpoints[0].input.val();
+    var input_to = endpoints[1].input.val();
+    var latlng_from = endpoints[0].latlng;
+    var latlng_to = endpoints[1].latlng;
+
+    endpoints[0].setLatLng(latlng_to);
+    endpoints[1].setLatLng(latlng_from);
+    endpoints[0].input.val(input_to);
+    endpoints[1].input.val(input_from);
+
+    $(".directions_form .routing_go").trigger("click");
+  });
```

Comment:
I think it would be better to do this in the same way we do it in other places, along these lines:

```
OSM.router.route("/directions?" + querystring.stringify({
  from: $("#route_to").val(),
  to: $("#route_from").val()
}));
```

Reply:
I originally did it that way, but it doesn't work when you paste a link with just coords in the to and from inputs.

It also jumps the marker to the nearest feature, rather than keeping it pin pointed in the exact coordinate (which leads to the problem OP mentioned about the true coord location not being retained). I'm not sure if it's just standard OSM behaviour to jump the marker to the nearest feature though.

Reply:
Basically .val() forces through an Geolocation API call which can mess with the actual user-submitted value or coord

Reply:
What about if you include a `route` parameter in the URL with the actual lat/lon values? I think in that case it will use those rather than rerunning the geocode? So something like:

```
OSM.router.route("/directions?" + querystring.stringify({
  from: $("#route_to").val(),
  to: $("#route_from").val(),
  route: from_lat + "," + from_lon + ";" + to_lat + "," + to_lon
}));
```

although I haven't looked at how to get hold of the lat/lon values yet.

Reply:
I've recommitted with some changes.

The OSM.router.route method fails to work in the example given in the linked Issue (I think its because the latlng values of each marker don't get set when linking straight from the URL, rather than after clicking the "Go" button)

Instead of making things too complicated, I've made a simple call to getRoute(), which handles the setting of all latlng values and includes a call to OSM.router.route.

It now works with the example in the linked Issue as well as normal situations when clicking the "Go" button